src/Features/Core/Portable/CodeStyle2/AbstractCodeStyle2Provider.cs

```diff
@@ -0,0 +1,352 @@
+                    // analyzer to run here.  However, we can still check to see if we could offer
+                    // the feature here as a refactoring.
+                    await ComputeRefactoringsAsync(context, optionValue.Value, forward: true).ConfigureAwait(false);
+                    return;
+
+                case ReportDiagnostic.Error:
+                case ReportDiagnostic.Warn:
+                case ReportDiagnostic.Info:
+                    // User has this option set at a level where we want it checked by the
+                    // DiagnosticAnalyser and not the CodeRefactoringProvider.  However, we still
+                    // want to check if we want to offer the *reverse* refactoring here in this
+                    // single location.
+                    //
+                    // For example, say this is the "use expression body" feature.  If the user says
+                    // they always prefer expression-bodies (with warning level), then we want the
+                    // analyzer to always be checking for that.  However, we still want to offer the
+                    // refactoring to flip their code to use a block body here, just in case that
+                    // was something they wanted to do as a one off (i.e. before adding new
+                    // statements.
+                    //
+                    // TODO(cyrusn): Should we only do this for warn/info?  Argument could be made
+                    // that we shouldn't even offer to refactor in the reverse direction if it will
+                    // just cause an error.  That said, maybe this is just an intermediary step, and
+                    // we shouldn't really be blocking the user from making it.
+                    await ComputeRefactoringsAsync(context, optionValue.Value, forward: false).ConfigureAwait(false);
+                    return;
+            }
+        }
+
+        private async Task ComputeRefactoringsAsync(
+            CodeRefactoringContext context, TOptionKind option, bool forward)
+        {
+            var document = context.Document;
+            var span = context.Span;
+            var cancellationToken = context.CancellationToken;
+
+            var computationTask = forward
+                ? ComputeForwardRefactoringsAsync(document, span, option, cancellationToken)
+                : ComputeReverseRefactoringsAsync(document, span, option, cancellationToken);
+
+            var codeActions = await computationTask.ConfigureAwait(false);
+            context.RegisterRefactorings(codeActions);
+        }
+
+        protected abstract Task<ImmutableArray<CodeAction>> ComputeForwardRefactoringsAsync(
+            Document document, TextSpan span, TOptionKind option, CancellationToken cancellationToken);
+        protected abstract Task<ImmutableArray<CodeAction>> ComputeReverseRefactoringsAsync(
+            Document document, TextSpan span, TOptionKind option, CancellationToken cancellationToken);
```

Comment:
this forces the subclass to realize that it needs to both offer the normal refactoring, as well as the reverse refactoring.